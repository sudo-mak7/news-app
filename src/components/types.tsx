import { CommentsInterface } from '@common/types/commentsInterface'

export interface AnswerSectionComponentInterface {
  isCollapsed: boolean,
  answers: CommentsInterface[],
  isLoading: boolean
}