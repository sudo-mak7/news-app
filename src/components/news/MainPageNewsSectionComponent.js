import React from 'react'
import {
  Header,
  Icon,
  Label,
  Segment
} from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { setCurrentNewsId } from '@redux/news/newsByIdSlice'
import { dateNormalizer } from '@utils/dateNormalizer'

const MainPageNewsSectionComponent = ({ title, by, score, time, id }) => {
  const dispatch = useDispatch()

  const date = dateNormalizer(time)

  const getNewsId = (e) => {
    localStorage.removeItem('currentNewsId')
    const id = e.currentTarget.id
    dispatch(setCurrentNewsId(id))
    localStorage.setItem('currentNewsId', JSON.stringify(id))
  }

  return (
    <Segment.Group
      id={ id }
      onClick={ (e) => getNewsId(e) }
    >
      <Link to='/news-page'>
        <Segment attached='top' color='blue'>
          <Header as='h2' color='black'>
            { title || 'No title' }
          </Header>
        </Segment>

        <Segment attached='bottom' clearing>
          <Label image floated='left'>
            <Icon name='user'></Icon>
            { by || 'No author' }
          </Label>
          <Label>
            <Icon name='star'></Icon>
              Rating
            <Label.Detail>
              { score || 'No rating' }
            </Label.Detail>
          </Label>
          <Label>
            <Icon name='calendar'></Icon>
            { date || 'No date' }
          </Label>
        </Segment>
      </Link>
    </Segment.Group>
  )
}

export default MainPageNewsSectionComponent