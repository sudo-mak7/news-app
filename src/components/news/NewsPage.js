import React from 'react'
import {
  Button,
  Container,
  Header,
  Icon,
  Label,
  Segment
} from 'semantic-ui-react'
import NewsText from './news-sections/NewsText'
import CommentsSection from '../comments/CommentsSection'

const NewsPage = ({ comment, comments }) => {
  return (
    <Container text style={{ marginTop: '5em' }}>
      <Segment.Group>
        <Segment attached='top' color='blue'>
          <Header as='h2' color='blue'>
            News' header
            <Button floated='right' color='blue'>Open a news source</Button>
          </Header>
        </Segment>

        <NewsText/>

        <Segment attached clearing>
          <Label image floated='left'>
            <Icon name='user'></Icon>
            Author
          </Label>
          <Label>
            <Icon name='comment'></Icon>
            Comments
            <Label.Detail>214</Label.Detail>
          </Label>
          <Label>
            <Icon name='calendar'></Icon>
            20.02.2022
          </Label>
        </Segment>

        <Segment attached='bottom'>
          {comments
            ? <CommentsSection comment={ comment }/>
            : <Header as='h4'>No comments</Header>
          }
        </Segment>

      </Segment.Group>
    </Container>
  )
}

export default NewsPage