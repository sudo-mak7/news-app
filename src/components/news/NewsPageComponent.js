import React, { useState } from 'react'
import {
  Button,
  Container,
  Header,
  Icon,
  Label,
  Segment
} from 'semantic-ui-react'
import CommentsSectionComponent from '../comments/CommentsSectionComponent'
import { dateNormalizer } from '../../utils/dateNormalizer'
import { useDispatch, useSelector } from 'react-redux'
import { setComments } from '../../redux/comments/commentsSlice'
import { loading } from '../../redux/loading/loadingSlice'
import { getComments } from '../../api/api'

const NewsPageComponent = () => {
  const news = useSelector(state =>
    (state.currentNews.currentNews || JSON.parse(localStorage.getItem('currentNews')))
    || []
  )


  const [ error, setError ] = useState('')

  const errorMessage = 'Error loading comments :('

  const dispatch = useDispatch()

  localStorage.setItem('currentNews', JSON.stringify(news))

  const updateComments = () => {
    if (news.hasOwnProperty('kids')) {
      dispatch(loading(true))

      getComments(
        news.kids
          ? news.kids
          : setError(errorMessage)
      ).then(data => {
        !data
          ? setError(errorMessage)
          : dispatch(setComments(data))
        dispatch(loading(false))
      })
    }
  }

  let date = ''

  if (typeof news.time === 'number') {
    date = dateNormalizer(news.time)
  }

  return (
    <Container text style={{ marginTop: '5em' }}>
      <Segment.Group>
        <Segment attached='top' color='blue'>
          <Header as='h2' color='blue'>
            { news.title || 'No title' }
          </Header>
        </Segment>

        <Segment attached clearing>
          <Label image floated='left'>
            <Icon name='user'></Icon>
            { news.by || 'No author' }
          </Label>
          <Label>
            <Icon name='comment'></Icon>
            Comments
            <Label.Detail>
              { news.descendants }
            </Label.Detail>
          </Label>
          <Label>
            <Icon name='calendar'></Icon>
            { date || 'No date' }
          </Label>

          { news.url
              ? <a href={ news.url }>
                  <Button size='mini' floated='right' color='blue'>
                    Open a news source
                  </Button>
                </a>
              : ''
          }

          <Button
            size='mini'
            floated='right'
            color='blue'
            onClick={ () => { updateComments() } }
          >
            Update comments
          </Button>

        </Segment>

        <Segment attached='bottom'>
          { news.hasOwnProperty('kids')
            ? <CommentsSectionComponent
                error={ error }
                setError={ setError }
              />
            : <Header as='h3' textAlign='center'>No comments yet</Header>
          }
        </Segment>

      </Segment.Group>
    </Container>
  )
}

export default NewsPageComponent