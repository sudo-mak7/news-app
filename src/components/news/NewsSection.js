import React from 'react'
import {
  Header,
  Icon,
  Label,
  Segment
} from 'semantic-ui-react'
import { Link } from 'react-router-dom'

const NewsSection = ({ title, by, score, time }) => {
  const getDate = new Date(time * 1000).toLocaleDateString() + ' ('
  const getHours = new Date(time * 1000).getHours() + ':'
  const getMinutes = new Date(time * 1000).getMinutes() < 10
    ? '0' + new Date(time * 1000).getMinutes() + ')'
    : new Date(time * 1000).getMinutes() + ')'

  const date = getDate + getHours + getMinutes

  return (
    <Segment.Group>
      <Link to='/news-page'>
        <Segment attached='top' color='blue'>
          <Header as='h2' color='black'>
            { title || 'No title' }
          </Header>
        </Segment>

        <Segment attached='bottom' clearing>
          <Label image floated='left'>
            <Icon name='user'></Icon>
            { by || 'No author' }
          </Label>
          <Label>
            <Icon name='star'></Icon>
              Rating
            <Label.Detail>
              { score || 'No rating' }
            </Label.Detail>
          </Label>
          <Label>
            <Icon name='calendar'></Icon>
            { date || 'No date' }
          </Label>
        </Segment>
      </Link>
    </Segment.Group>
  )
}

export default NewsSection