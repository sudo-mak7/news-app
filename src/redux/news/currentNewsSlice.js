import { createSlice } from '@reduxjs/toolkit'

export const currentNewsSlice = createSlice({
  name: 'currentNews',
  initialState: {
    currentNews: JSON.parse(localStorage.getItem('currentNews'))
  },
  reducers: {
    setCurrentNews: (state, action) => {
      state.currentNews = action?.payload
    }
  }
})

export const { setCurrentNews } = currentNewsSlice.actions

export default currentNewsSlice.reducer