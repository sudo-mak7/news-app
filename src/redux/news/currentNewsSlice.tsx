import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { getItemFromApiWithUrlById } from '@api/url'
import { fetchComments } from '@redux/comments/commentsSlice'
import { NewsInterface } from '@common/types/newsInterface'
import { CurrentNewsStateInterface } from '@redux/types'
import {ERROR_LOADING_CURRENT_NEWS_MESSAGE} from "@common/messages/errors";

export const fetchCurrentNewsById = createAsyncThunk(
  'currentNews/fetchCurrentNewsById',
  async (id: number, { rejectWithValue , dispatch }) => {
    try {
      const response = await fetch(getItemFromApiWithUrlById(id))

      if (!response.ok) {
        console.warn(ERROR_LOADING_CURRENT_NEWS_MESSAGE)
        throw new Error(ERROR_LOADING_CURRENT_NEWS_MESSAGE)
      } else {
        const currNews = await response.json()
        dispatch(currentNews(currNews))
        dispatch(fetchComments(currNews.kids))
      }
    } catch (error: any) {
      return rejectWithValue(error.message)
    }
  }
)

const initialState = {
  currentNews: {},
  loading: false,
  error: null
} as CurrentNewsStateInterface

export const currentNewsSlice = createSlice({
  name: 'currentNews',
  initialState,
  reducers: {
    currentNews: (state, action: PayloadAction<NewsInterface>) => {
      state.currentNews = action.payload
    }
  },
  extraReducers: (builder) => {
    builder.addCase(fetchCurrentNewsById.pending, (state) => {
      state.loading = true
      state.error = null
    })
      .addCase(fetchCurrentNewsById.fulfilled, (state) => {
        state.loading = false
      })
      .addCase(fetchCurrentNewsById.rejected, (state, action: PayloadAction<any>) => {
        state.loading = false
        state.error = action.payload
      })
  }
})

export const { currentNews } = currentNewsSlice.actions

export default currentNewsSlice.reducer