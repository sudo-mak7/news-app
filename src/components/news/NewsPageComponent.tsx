import * as React from 'react'
import { useEffect } from 'react'
import {
  Button,
  Container,
  Header,
  Icon,
  Label,
  Loader,
  Segment
} from 'semantic-ui-react'
import { useAppDispatch } from '@redux/reduxHooks'
import { useAppSelector } from '@redux/reduxHooks'
import { fetchCurrentNewsById } from '@redux/news/currentNewsSlice'
import { fetchComments } from '@redux/comments/commentsSlice'
import {
  getCurrentNewsErrorSelector,
  getCurrentNewsLoaderSelector,
  getCurrentNewsSelector,
  getNewsByIdSelector
} from '@redux/selectors'
import { windowScrollUp } from '@utils/windowScrollUp'
import { dateNormalizer } from '@utils/dateNormalizer'
import CommentsSectionComponent from '@components/comments/CommentsSectionComponent'

const NewsPageComponent = (): JSX.Element => {
  const newsId = useAppSelector(getNewsByIdSelector)
  const news = useAppSelector(getCurrentNewsSelector)
  const isLoading = useAppSelector(getCurrentNewsLoaderSelector)
  const error = useAppSelector(getCurrentNewsErrorSelector)

  const errorMessage = 'Error loading current news :('

  const dispatch = useAppDispatch()

  useEffect((): void => {
    windowScrollUp()
    dispatch(fetchCurrentNewsById(newsId))
  }, [])

  const updateComments = (): void => {
    if (news.hasOwnProperty('kids')) {
      dispatch(fetchComments(news.kids))
    }
  }

  let date = ''

  if (typeof news.time === 'number') {
    date = dateNormalizer(news.time)
  }

  return (
    <Container text style={{ marginTop: '5em' }}>
      { isLoading
          ? <Loader active/>
          : error
              ? <Header as='h2' color='blue' textAlign='center'>{ errorMessage }</Header>
              : <Segment.Group>
                  <Segment attached='top' color='blue'>
                    <Header as='h2' color='blue'>
                      { news.title || 'No title' }
                    </Header>
                  </Segment>

                  <Segment attached clearing>
                    <Label image floated='left'>
                      <Icon name='user'></Icon>
                      { news.by || 'No author' }
                    </Label>
                    <Label>
                      <Icon name='comment'></Icon>
                      Comments
                      <Label.Detail>
                        { news.descendants }
                      </Label.Detail>
                    </Label>
                    <Label>
                      <Icon name='calendar'></Icon>
                      { date || 'No date' }
                    </Label>

                    { news.url
                      ? <a href={ news.url }>
                          <Button size='mini' floated='right' color='blue'>
                            Open a news source
                          </Button>
                        </a>
                      : ''
                    }

                    <Button
                      size='mini'
                      floated='right'
                      color='blue'
                      onClick={ (): void => { updateComments() } }
                    >
                      Update comments
                    </Button>

                  </Segment>

                  <Segment attached='bottom'>
                    { news.hasOwnProperty('kids')
                        ? <CommentsSectionComponent />
                        : <Header as='h3' textAlign='center'>No comments yet</Header>
                    }
                  </Segment>

                </Segment.Group>
      }
    </Container>
  )
}

export default NewsPageComponent