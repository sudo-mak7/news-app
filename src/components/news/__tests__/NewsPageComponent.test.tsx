import * as React from 'react'
import * as reduxHooks from '@redux/reduxHooks'
import { render, screen, fireEvent } from '@testing-library/react'
import { Provider } from 'react-redux'
import { store } from '@redux/store'
import { currentNewsSlice, fetchCurrentNewsById } from '@redux/news/currentNewsSlice'
import NewsPageComponent from '@components/news/NewsPageComponent'
import { CurrentNewsStateInterface } from '@redux/types'

const initialState = {
  currentNews: {},
  loading: false,
  error: null
} as CurrentNewsStateInterface

const currentNewsSliceReducer = currentNewsSlice.reducer

const newsPageComponent = (): JSX.Element =>
  <Provider store = { store } >
    <NewsPageComponent />
  </Provider>

const mockedDispatch = jest.spyOn(reduxHooks, 'useAppDispatch')

describe('NewsPageComponent test', () => {
  test('NewsPageComponent should to match snapshot', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(newsPageComponent())
    expect(newsPageComponent()).toMatchSnapshot()
  })

  test('renders NewsPageComponent update comments button with completed loading', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    const state = currentNewsSliceReducer(initialState, fetchCurrentNewsById.fulfilled)
    expect(state.loading).toBe(false)

    render(newsPageComponent())

    const buttonInNewsPageComponent = screen.getByRole('button')
    expect(buttonInNewsPageComponent).toHaveTextContent(/update comments/i)
  })

  test('renders NewsPageComponent with no data', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    const state = currentNewsSliceReducer(initialState, fetchCurrentNewsById.fulfilled)
    expect(state.loading).toBe(false)

    render(newsPageComponent())

    const newsInNewsPageComponent = screen.getByTestId('news')
    expect(newsInNewsPageComponent).toBeInTheDocument()

    const titleInNewsPageComponent = screen.getByText(/no title/i)
    expect(titleInNewsPageComponent).toBeInTheDocument()

    const authorInNewsPageComponent = screen.getByText(/no author/i)
    expect(authorInNewsPageComponent).toBeInTheDocument()

    const commentsInNewsPageComponent = screen.getByText(/no comments yet/i)
    expect(commentsInNewsPageComponent).toBeInTheDocument()

    const dateInNewsPageComponent = screen.getByText(/no date/i)
    expect(dateInNewsPageComponent).toBeInTheDocument()
  })

  test('dispatch should called once by click', () => {
    window.scrollTo = jest.fn()

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(newsPageComponent())
    const buttonInNewsPageComponent = screen.getByRole('button')

    fireEvent.click(buttonInNewsPageComponent)
    expect(dispatch).toHaveBeenCalledTimes(1)
  })

  afterAll(() => {
    jest.clearAllMocks()
  })
})
