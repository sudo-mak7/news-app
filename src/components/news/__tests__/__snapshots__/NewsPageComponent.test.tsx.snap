// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsPageComponent test NewsPageComponent should to match snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <NewsPageComponent />
</Provider>
`;

exports[`NewsPageComponent test renders NewsPageComponent loader 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <NewsPageComponent />
</Provider>
`;
